File: examples/kotlin/nats-stream-client/src/test/kotlin/cdcexamples/natsstream/FormatTest.kt
package cdcexamples.natsstream

import com.google.gson.JsonParser
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Prüft [Format.formatChange] — reine Funktion, netzlos testbar, inklusive
 * des `DELETE`-Falls, in dem das Row Image auf dem Draht das JSON-Literal
 * `null` trägt (dieselbe Übersetzung wie `natsstream.rowImage` auf der
 * Erzeugerseite). Form-Vorbild:
 * `examples/kotlin/grpc-client/src/test/kotlin/cdcexamples/grpc/FormatTest.kt`.
 */
class FormatTest {
    private fun json(raw: String) = JsonParser.parseString(raw)

    @Test
    fun formatChangeCarriesIdentityAndPayload() {
        val change = StreamMessage(
            changeId = "c-1",
            transactionId = "t-1",
            sourceTableId = "st-1",
            sequence = 1,
            operation = "INSERT",
            oldImage = json("null"),
            newImage = json("{\"id\":1}"),
            schemaVersion = "v1",
            schema = "public",
            table = "orders",
        )

        val got = Format.formatChange(change)

        assertEquals(
            "nats-stream-client: change_id=c-1 table=public.orders operation=INSERT new_image={\"id\":1}",
            got,
        )
    }

    @Test
    fun formatChangeHandlesMissingNewImage() {
        val change = StreamMessage(
            changeId = "c-2",
            transactionId = "t-1",
            sourceTableId = "st-1",
            sequence = 2,
            operation = "DELETE",
            oldImage = json("{\"id\":1}"),
            newImage = json("null"),
            schemaVersion = "v1",
            schema = "public",
            table = "orders",
        )

        val got = Format.formatChange(change)

        assertEquals(
            "nats-stream-client: change_id=c-2 table=public.orders operation=DELETE new_image=null",
            got,
        )
    }
}

File: examples/nats-stream-client/format_test.go
package main

import (
	"encoding/json"
	"testing"
)

// TestFormatChangeCarriesIdentityAndPayload prüft den Aufbau der Ausgabezeile
// (LH-FA-SST-008): Tabelle, Operation und change_id stehen darin, das neue
// Row Image wird unverändert angehängt.
func TestFormatChangeCarriesIdentityAndPayload(t *testing.T) {
	c := streamMessage{
		ChangeID:  "c-1",
		Schema:    "public",
		Table:     "orders",
		Operation: "INSERT",
		NewImage:  []byte(`{"id":1}`),
	}
	got := formatChange(c)
	want := `nats-stream-client: change_id=c-1 table=public.orders operation=INSERT new_image={"id":1}`
	if got != want {
		t.Fatalf("formatChange = %q, want %q", got, want)
	}
}

// TestFormatChangeHandlesMissingNewImage prüft den DELETE-Fall (SPEC-024):
// ein fehlendes Row Image trägt auf dem Draht das JSON-Literal `null`
// (dieselbe Übersetzung wie `natsstream.rowImage` auf der Erzeugerseite) —
// nach dem Dekodieren stehen genau diese vier Bytes im Feld, kein leerer
// Wert.
func TestFormatChangeHandlesMissingNewImage(t *testing.T) {
	c := streamMessage{
		ChangeID:  "c-2",
		Schema:    "public",
		Table:     "orders",
		Operation: "DELETE",
		NewImage:  json.RawMessage("null"),
	}
	got := formatChange(c)
	want := `nats-stream-client: change_id=c-2 table=public.orders operation=DELETE new_image=null`
	if got != want {
		t.Fatalf("formatChange = %q, want %q", got, want)
	}
}
